OptimizelySDK.Tests/OptimizelyTest.cs

```diff
@@ -3191,8 +3194,185 @@ public void TestGetFeatureVariableDoubleSendsNotificationWhenUserNotBuckedIntoBo
+            Assert.Null(result);
+
+            LoggerMock.Verify(log => log.Log(LogLevel.WARN, "The featureKey parameter must be nonnull."), Times.Once);
+        }
+
+        [Test]
+        public void TestGetAllFeatureVariablesNullUserID()
+        {
+            var userAttributes = new UserAttributes
+            {
+               { "device_type", "iPhone" },
+               { "company", "Optimizely" },
+               { "location", "San Francisco" }
+            };
+            var optly = Helper.CreatePrivateOptimizely();
+
+            var result = (OptimizelyJson)optly.Invoke("GetAllFeatureVariables", "string_single_variable_feature", null, userAttributes);
+            Assert.Null(result);
+
+            LoggerMock.Verify(log => log.Log(LogLevel.WARN, "The userId parameter must be nonnull."), Times.Once);
+        }
+        
+        [Test]
+        public void TestGetAllFeatureVariablesEmptyOrNullFeatureFlag()
+        {
+            var featureKey = "InvalidFeatureKey";
+            var userAttributes = new UserAttributes
+            {
+               { "device_type", "iPhone" },
+               { "company", "Optimizely" },
+               { "location", "San Francisco" }
+            };
+            var optly = Helper.CreatePrivateOptimizely();
+
+            var result = (OptimizelyJson)optly.Invoke("GetAllFeatureVariables", featureKey, TestUserId, userAttributes);
+            Assert.Null(result);
+
+            LoggerMock.Verify(log => log.Log(LogLevel.INFO, "No feature flag was found for key \"" + featureKey + "\"."), Times.Once);
+        }
+        
+        [Test]
+        public void TestGetAllFeatureVariablesInvalidDatafile()
+        {
+            var userAttributes = new UserAttributes
+            {
+               { "device_type", "iPhone" },
+               { "company", "Optimizely" },
+               { "location", "San Francisco" }
+            };
+            var optly = new Optimizely("Random datafile", null, LoggerMock.Object);
+
+            var result = optly.GetAllFeatureVariables("validFeatureKey", TestUserId, userAttributes);
+            
+            Assert.Null(result);
+
+            LoggerMock.Verify(log => log.Log(LogLevel.ERROR, "Optimizely instance is not valid, failing getAllFeatureVariableValues call. type"), Times.Once);
+        }
+
+        [Test]
+        public void TestGetAllFeatureVariablesRollout()
+        {
+            var featureKey = "string_single_variable_feature";
+            var experiment = Config.GetRolloutFromId("166661").Experiments[0];
+
+            var userAttributes = new UserAttributes
+            {
+               { "device_type", "iPhone" },
+               { "company", "Optimizely" },
+               { "location", "San Francisco" }
+            };
+
+            var featureFlag = Config.GetFeatureFlagFromKey(featureKey);
+
+            var decision = new FeatureDecision(experiment, null, FeatureDecision.DECISION_SOURCE_ROLLOUT);
```

Comment:
Why variation is null.

Reply:
To verify info log that ""User \"" + TestUserId + "\" was not bucketed into any variation for feature flag \"" + featureKey + "\". The default values are being returned.""